sql/parser/eval.go

```diff
@@ -134,15 +137,23 @@ var binOps = map[binArgs]binOp{
 		},
 	},
 	binArgs{Plus, dateType, intervalType}: {
-		returnType: DummyTimestamp,
+		returnType: DummyDate,
 		fn: func(left Datum, right Datum) (Datum, error) {
-			return DTimestamp{Time: left.(DDate).Add(right.(DInterval).Duration)}, nil
+			days, err := inDays(right.(DInterval).Duration)
+			if err != nil {
+				return DummyDate, err
+			}
+			return DDate(left.(DDate) + days), nil
```

Comment:
Ah, that makes more sense, though (sadly) I believe we're going to have to revisit our stance on implicit casts as no other system is as strict as we are.